### PR TITLE
NIST Support

### LINKED ISSUE #862: [FEATURE] Embed new file format parser for NIST file format
Please add requirements on this issue @collinsr

### PR DESCRIPTION
**Pull request by BryonLewis**
Description:
NIST format has the concepts of activities which are typically a full frame action that is happening.  Some of the specifications I have looked at also include a concept of Objects, which these are individual persons/things being tracks at a bounding box level during an Activity. 

Link to specification: https://actev.nist.gov/pub/Phase3_ActEV_2021_SDL_EvaluationPlan_20210803.pdf

Notes:

- Since activities are temporal and not spatial I opted to have them align themselves on the left side of the screen with space between concurrent activities.
- NIST and our track files are both JSON so similar to the kwcoco on the web side I need to load the file up initiall if it is JSON and do some validation to make sure it is a NIST file and not a track.json file.  
- I added support for exporting and importing NIST files into the diveCLI.
- Added some testing procedures for my current implementation of the NIST format.

fixes #862

### CODE REVIEW

File: client/platform/desktop/backend/native/common.ts [hunk 1]
```diff
@@ -827,10 +845,21 @@ async function _importTrackFile(
...
+          const track = trackData.tracks[i];
+          trackStructure[track.trackId] = track;
+        }
+        const serialized = JSON.stringify(trackStructure);
+        await fs.writeFile(newPath, serialized);
+      } else {
+        await fs.copy(
+          trackFileAbsPath,
+          newPath,
+        );
```
Comment (subdavis) [inline client/platform/desktop/backend/native/common.ts:861]: Looks like copy/paste from just a few lines up.  Why are `_importTrackFile` and `annotationImport` different functions?

File: client/platform/desktop/constants.ts [hunk 2]
```diff
@@ -213,6 +215,7 @@ export interface CheckMediaResults {
   websafe: boolean;
   originalFpsString: string;
   originalFps: number;
+  videoDimensions: {width: number; height: number};
```
Comment (subdavis) [inline client/platform/desktop/constants.ts:218]: what's going on here?  Is this related to nist?

**Review by subdavis — COMMENTED** [review]
I don't know when to address this, but I've officially lost all my bearings on common.ts

I think we need to unify the code that runs import at dataset creation time and the code that runs overwrite import.  

* `annotationImport`
* `_importTrackFile`
* `findTrackFileinFolder`
* `processOtherAnnotationFiles`
* `processAnnotationFilePath`

File: client/platform/desktop/constants.ts (hunk 2 above)
Comment (BryonLewis) [reply client/platform/desktop/constants.ts:218]: Multiple activities have no spatial information and all tracks have associated bounding boxes.  This is used to organize the NIST activities on the screen in a row on the left side of the screen so you can see them visually.  Was going to do full frame but then overlapping activities would be unreadable in the viewer.

File: client/platform/desktop/backend/native/common.ts (hunk 1 above)
Comment (BryonLewis) [reply client/platform/desktop/backend/native/common.ts:861]: There may be a refactor that can combine them maybe with more arguments.  `_importTrackFile` is an initial import when loading a dataset so it is moving data around to place the `result_` in the proper location while extracting and updating jsonMeta for the import.

`annotationImport` is importing data into an already existing dataset so it acquires a lock and only places the file in the proper location with `result_` naming scheme so it will auto-load on refresh.

I'm just writing notes out below for myself when refactoring to combine the duplicate code

`_importTrackFile`:
- calculates and saves Metadata for loading
- Creates empty file if no file is specified
- Returns metadata for adding to localStorage

`annotationImport`:
- Determines if the specified file is valid track.json/NIST/csv
- acquires lock on existing track file
- deletes existing track file
- creates JSON from track file and moves/renames it so it can be autoloaded
- returns true to indicate successful import.

The easy thing to combine is the checking/generation of the track.json file that is used into something like `generateTrackJson` which will take a file path and either generate a track.json or return empty if it cna't find a valid format.,

File: client/platform/desktop/backend/native/common.ts [hunk 3]
```diff
@@ -830,10 +848,21 @@ async function _importTrackFile(
 
     const newPath = npath.join(projectDirAbsPath, npath.basename(newFileName));
 
-    await fs.copy(
-      userTrackFileAbsPath,
-      newPath,
-    );
+    if (jsonMeta.originalVideoFile && await nistSerializers.confirmNistFile(userTrackFileAbsPath)) {
```
Comment (subdavis) [inline client/platform/desktop/backend/native/common.ts:851]: I'm confused by this line.  isnt `type` more appropriate than `originalVideoFile` for determining dataset type?

I also kinda don't like that we have a data type that only applies to video.   In the `annotationImport` function above, it doesn't look like you've added any code to enforce that the user must only import NIST data on videos.

File: client/platform/desktop/backend/native/common.ts [hunk 4]
```diff
@@ -830,10 +848,21 @@ async function _importTrackFile(
...
+      const trackData = await nistSerializers.loadNistFile(userTrackFileAbsPath);
```
Comment (subdavis) [inline client/platform/desktop/backend/native/common.ts:852]: Where is `fullFrameBounds` passed?  Looks like nothing uses that argument.

File: client/platform/desktop/backend/serializers/nist.ts [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+    localization: ActivityLocalization;
+    activityID: number;
+    objects?: NistObject[];
+}
+
+function confirmNistFormat(data: NistFile) {
+  return data.activities && data.filesProcessed;
+}
+
+async function confirmNistFile(filename: string) {
```
Comment (subdavis) [inline client/platform/desktop/backend/serializers/nist.ts:69]: return value of this function is weird.  Callers all treat it like a boolean.  Could the return type be a proper bool rather than `string[] | { tracks: never[]; }`?

**Review by subdavis — CHANGES_REQUESTED** [review]

File: client/platform/desktop/backend/native/common.ts (hunk 4 above)
Comment (BryonLewis) [reply client/platform/desktop/backend/native/common.ts:852]: The CLI version uses it when doing conversions, in all other instances it defaults to 1080p-ish size.  I was hesitant about making the system do another `checkMedia` call on import for each NIST file or storing it in the metadata.  For right now I'm going to default to the full frame size given that the source NIST videos typically are that resolution.

File: client/platform/desktop/backend/native/common.ts (hunk 3 above)
Comment (BryonLewis) [reply client/platform/desktop/backend/native/common.ts:851]: I was just looking at it and while I can enforce the type check on the `annotationImport` I think the right thing to do is to throw Proper error messages in the annotation import.  `annotationImport.vue` right now just looks for simple boolean value on import I think that should change on the desktop/server end to be wrapped in a try/catch and catch and display errors as to why it failed.  That might be better done in a different PR.

Comment (subdavis) [reply client/platform/desktop/backend/native/common.ts:851]: Sorry I'm not following.

It seems to me like this line should not use `jsonMeta.originalVideoFile`.  Is there a reason you used that as a proxy for `jsonMeta.type`, and can it be changed?

Comment (BryonLewis) [reply client/platform/desktop/backend/native/common.ts:851]: Yeah I'm sorry I wasn't clear there and jumped a few steps ahead.  The `originalVideoFile` was a mistake on my part.  I fixed that to use `jsonMeta.type`.  The second part of your comment about `annotationImport` not having any guards against applying on video I fixed that as well using the `jsonMeta.type` to prevent NIST loading on image-sequences. 

The `annotationImport` function returns `true` for imported and `false` for rejected.  When rejected it just displays a generic prompt saying "failed to import {X}".  I'm saying that this should be refactored on both the desktop/server side to throw descriptive errors and the client should display the error.  That way they can see that they can't import a NIST file on image-sequences.

**Comment by BryonLewis** [comment]
We may need some discussion about the `fullFrameBounds` and if it should always read the media to figure out the video size or if I should modify the metadata to store this information.  This is a thing unique to NIST in that there are activities (tracks) not associated with any bounds so I need to generate them.

File: client/platform/desktop/backend/serializers/nist.ts [hunk 6]
```diff
@@ -63,14 +63,14 @@ interface NistActivity {
 }
 
 function confirmNistFormat(data: NistFile) {
-  return data.activities && data.filesProcessed;
+  return !!(data.activities && data.filesProcessed);
```
Comment (subdavis) [inline client/platform/desktop/backend/serializers/nist.ts:66]: This is not equivalent.

`!!([]) === true` even though `[] == false`.

It's best not to rely on squishy truthiness and do strict null, undefined, or length assertions.

File: client/platform/desktop/backend/serializers/nist.ts [hunk 7]
```diff
@@ -63,7 +63,7 @@ interface NistActivity {
 }
 
 function confirmNistFormat(data: NistFile) {
-  return data.activities !== undefined && data.filesProcessed !== undefined;
+  return Array.isArray(data.activities) && Array.isArray(data.filesProcessed);
```
Comment (subdavis) [inline client/platform/desktop/backend/serializers/nist.ts:66]: Either the interface is incorrect or this will never return false because those files can only be an array.

https://github.com/Kitware/dive/pull/883/files#diff-9179ddbd5a97d67ddf6d93b78f1534efec3cb8f44e3070606a5c0aad890666bcR38-R49

Comment (BryonLewis) [reply client/platform/desktop/backend/serializers/nist.ts:66]: https://github.com/Kitware/dive/pull/883/files#diff-9179ddbd5a97d67ddf6d93b78f1534efec3cb8f44e3070606a5c0aad890666bcR76

https://github.com/Kitware/dive/pull/883/files#diff-9179ddbd5a97d67ddf6d93b78f1534efec3cb8f44e3070606a5c0aad890666bcR271

I'm doing this backwards in that I should restructure the confirmNist to handle `any` or some sort of JSON object, confirm that it is a NISTFile and then rely on that to be the Type guard for the `as NistFile` assignment

Comment (subdavis) [reply client/platform/desktop/backend/serializers/nist.ts:66]: Oh, I see.  Please use a type guard instead.   The parameter to `confirmNistFormat` should be `any` if you don't know what it is going in. 

```
function isNistFile(data: any): data is NistFile {
  return Array.isArray(data.activities) && Array.isArray(data.filesProcessed);
}
```

then, inside a block where `isNistFile(data)` is true, typescript knows `data` is of type `NistFile`

**Review by subdavis — APPROVED** [review]
I sure do love making things harder than they need to be :)

**Comment by BryonLewis** [comment]
> I sure do love making things harder than they need to be :)

It wasn't you, I've been kind of half jumping back and forth on this and I have this internal idea of how TS should work and it doesn't.  There should really be an exception to `any` in the linter for type guards.  I just looked at `any` as being a forbidden thing and got tunneled into my very bad logic there.

**Comment by subdavis** [comment]
> There should really be an exception to any in the linter for type guards.

Yes, there should be.
